Add build support for kineto + rocm

mwootton opened the pull request:
Fixes #58399

CMake changes to allow kineto to build with rocm support.

mwootton committed:
Add build support for kineto + rocm

mwootton committed:
Fix whitespace

walterddr commented on cmake/Dependencies.cmake:
```diff
@@ -1828,10 +1828,25 @@ endif()
+      set(LIBKINETO_NOROCTRACER OFF CACHE STRING "")
+      message(STATUS "Using Kineto with Roctracer support")
+    endif()
+    if(USE_CUDA)
+      set(LIBKINETO_NOCUPTI OFF CACHE STRING "")
+      message(STATUS "Using Kineto with CUPTI support")
+    endif()
```
This part seems wrong if `MSVC` is set to true and `USE_CUDA` is set to true

mwootton replied:
Sorry, it is a bit convoluted.  I tried to make minimal changes to something that clearly only considers one tracing library.
There is also a hidden relation: USE_CUDA && USE_ROCM == False.  These never coexist.

So the 1846 guard can only be reached if 1829 is not entered.  If MSVC is set then LIBKINETO_NOCUPTI is ON.

At that point the only way into 1838 is if we are a ROCm build.  So then the 1846 is always false.

Also, there is no rocm on windows (i.e. msvc).

Happy to refactor it.  Suggestions welcome.

walterddr replied:
Hmm I guess here it is a bit confusing to me because 
both `LIBKINETO_NOCUPTI` and `LIBKINETO_NOROCTRACER` are cached strings. therefore users can override it. and the set statement here are noops if user has explicitly set cache.

walterddr reviewed (commented):
the if else logic here seems convoluted. 
Can we separate the logic setting
`LIBKINETO_NOROCTRACER`
and 
`LIBKINETO_NOROCTRACER`?

Seems the only part that they need to be used together is printing the message "cpu-only".

mwootton commented:
Indeed, can get rid of the "else" and replace with:
```
if(USE_ROCM AND NOT LIBKINETO_NOROCTRACER)
endif()
if(USE_CUDA AND NOT MSVC AND NOT LIBKINETO_NOCUPTI)
endif()
```
Embrace the double negatives.  Seems easier to follow.  Agree?

mwootton commented:
Or this:
```
  if((NOT USE_CUDA) OR MSVC)
    set(LIBKINETO_NOCUPTI ON CACHE STRING "")
  elseif(USE_CUDA)  # MSVC is FALSE
    set(LIBKINETO_NOCUPTI OFF CACHE STRING "")
    message(STATUS "Using Kineto with CUPTI support")
  endif()

  if(NOT USE_ROCM)
    set(LIBKINETO_NOROCTRACER ON CACHE STRING "")
  else()
    if(NOT ENV{ROCM_SOURCE_DIR})
      set(ENV{ROCM_SOURCE_DIR} "/opt/rocm")
    endif()
    set(LIBKINETO_NOROCTRACER OFF CACHE STRING "")
    message(STATUS "Using Kineto with Roctracer support")
  endif()

  if(LIBKINETO_NOCUPTI AND LIBKINETO_NOROCTRACER)
    message(STATUS "Using CPU-only version of Kineto")
  endif()
```

walterddr commented on cmake/Dependencies.cmake:
```diff
@@ -1828,10 +1828,25 @@ endif()
 if(USE_KINETO)
   if((NOT USE_CUDA) OR MSVC)
     set(LIBKINETO_NOCUPTI ON CACHE STRING "")
+  endif()
```
```suggestion
  else()
    set(LIBKINETO_NOCUPTI OFF CACHE STRING "")
  endif()
```

walterddr commented on cmake/Dependencies.cmake:
```diff
@@ -1828,10 +1828,25 @@ endif()
+  endif()
+  if(NOT USE_ROCM)
+    set(LIBKINETO_NOROCTRACER ON CACHE STRING "")
+  endif()
```
```suggestion
  else()
    set(LIBKINETO_NOROCTRACER OFF CACHE STRING "")
  endif()
```

walterddr commented on cmake/Dependencies.cmake:
```diff
@@ -1828,10 +1828,25 @@ endif()
+
+  if(LIBKINETO_NOCUPTI AND LIBKINETO_NOROCTRACER)
     message(STATUS "Using CPU-only version of Kineto")
   else()
-    set(LIBKINETO_NOCUPTI OFF CACHE STRING "")
-    message(STATUS "Using Kineto with CUPTI support")
+    if(USE_ROCM)
+      if(NOT ENV{ROCM_SOURCE_DIR})
+        set(ENV{ROCM_SOURCE_DIR} "/opt/rocm")
+      endif()
+      set(LIBKINETO_NOROCTRACER OFF CACHE STRING "")
+      message(STATUS "Using Kineto with Roctracer support")
+    endif()
+    if(USE_CUDA)
+      set(LIBKINETO_NOCUPTI OFF CACHE STRING "")
+      message(STATUS "Using Kineto with CUPTI support")
+    endif()
```
```suggestion
# Only one tracing library can be used. 
  if(NOT LIBKINETO_NOCUPTI)
    set(LIBKINETO_NOROCTRACER ON CACHE STRING "" FORCE)
    if(NOT USE_CUDA)
      message(WARNING "Kineto tracing requested CUPIT but USE_CUDA not set)
      set(LIBKINETO_NOCUPTI ON CACHE STRING "" FORCE)
    else()
      message(STATUS "Using Kineto with CUPTI support")
    endif()
  elif(LIBKINETO_NOROCTRACER)
    set(LIBKINETO_NOCUPTI ON CACHE STRING "" FORCE)
    if(NOT USE_CUDA)
      message(WARNING "Kineto tracing requested ROCTRACER but USE_ROCM not set)
      set(LIBKINETO_NOROCTRACER ON CACHE STRING "" FORCE)
    else()
      message(STATUS "Using Kineto with Roctracer support")
    endif()
  else()
    message(STATUS "Using CPU-only version of Kineto")
  endif()
```

mwootton committed:
Revert "Fix whitespace"

This reverts commit 87a7508e05b7d3a54f3b4e98b5769053345629f0.

mwootton committed:
Revert "Add build support for kineto + rocm"

This reverts commit 6421d419fe44ce096720fd4ec7482e65517af72a.

mwootton committed:
Add build support for kineto + rocm

mwootton commented:
I think I have this adequately cleaned up now.

1. There is already logic that prevents USE_CUDA and USE_ROCM is the same pytorch build, no need to duplicate it
2. **Change**: If USE_CUDA is off we should FORCE disable cupti.
3. I added cloned logic to set up KINETO_NOROCTRACER
4. I added a cloned block to set up provisions for roctracer

mwootton commented:
Thank you.
